### PR TITLE
Move ReadFromEnum and CreateEnumerable and add sample

### LINKED ISSUE #2609: Do something to ComponentCreation.ReadFromEnumerable
There's this little ReadFromEnumerable method inside ComponentCreation that is the only remaining public method. While the method is useful, it doesn't really seem like it belongs there, but rather as perhaps either some method directly on the data catalog, or somewhere more appropriately named.

Once that is moved somewhere else the class itself can be made best-friend-internal, all methods inside made public.

Based on a comment from @Ivanidzo4ka in #2605.

### PR DESCRIPTION
**Pull request by artidoro**
Fixes #2609.

1. I moved `ReadFromEnumerable` method outside of `ComponentCreation`. I could therefore make the class `ComponentCreation` BestFriend internal, while making the internal methods public. 
2. I moved the extension method for `CreateEnumerable` which used to be under `MLContext` to the `Data` extensions where `ReadFromEnumerable` lives. I could therefore make the `CursoringUtils` class internal.
3. I added a sample for `CreateEnumerable` and `ReadFromEnumerable`.

Only non trivial changes are in DataViewEnumerable.cs, DataOperationsCatalog.cs, MLContext.cs, TypedCursor.cs, ComponentCreation.cs. The rest are simply adjusting the code to use `mlContext.Data.CreateEnumerable`.

### CODE REVIEW

File: src/Microsoft.ML.Data/Utilities/ComponentCreation.cs [hunk 1]
```diff
@@ -15,7 +15,8 @@ namespace Microsoft.ML.Data
     /// This class defines extension methods for an <see cref="IHostEnvironment"/> to facilitate creating
     /// components (loaders, transforms, trainers, scorers, evaluators, savers).
     /// </summary>
-    public static class ComponentCreation
+    [BestFriend]
+    internal static class ComponentCreation
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/Utilities/ComponentCreation.cs]: >ComponentCreation [](start = 26, length = 17)

For those who wonder
This is main change A.

File: src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs [hunk 2]
```diff
@@ -24,6 +25,65 @@ internal DataOperationsCatalog(IHostEnvironment env)
             Environment = env;
         }
 
+        /// <summary>
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs]: >summary [](start = 13, length = 7)

And this is main change B

File: src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs [hunk 3]
```diff
@@ -24,6 +25,65 @@ internal DataOperationsCatalog(IHostEnvironment env)
...
+        /// Create a new <see cref="IDataView"/> over an enumerable of the items of user-defined type.
+        /// The user maintains ownership of the <paramref name="data"/> and the resulting data view will
+        /// never alter the contents of the <paramref name="data"/>.
+        /// Since <see cref="IDataView"/> is assumed to be immutable, the user is expected to support
+        /// multiple enumeration of the <paramref name="data"/> that would return the same results, unless
+        /// the user knows that the data will only be cursored once.
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs]: >cursored [](start = 54, length = 8)

enumerated?

File: test/Microsoft.ML.Tests/Scenarios/Api/CookbookSamples/CookbookSamples.cs [hunk 4]
```diff
@@ -66,7 +66,7 @@ private void IntermediateData(string dataPath)
             // 'transformedData' is a 'promise' of data. Let's actually read it.
             var someRows = mlContext
                 // Convert to an enumerable of user-defined type. 
-                .CreateEnumerable<InspectedRow>(transformedData.AsDynamic, reuseRowObject: false)
+                .Data.CreateEnumerable<InspectedRow>(transformedData.AsDynamic, reuseRowObject: false)
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Tests/Scenarios/Api/CookbookSamples/CookbookSamples.cs]: >CreateEnumerable [](start = 22, length = 16)

don't forget to update cook book files!

Comment (artidoro) [reply test/Microsoft.ML.Tests/Scenarios/Api/CookbookSamples/CookbookSamples.cs]: Good point thank you!

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/DataViewEnumerable.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+            var mlContext = new MLContext();
+
+            // Get a small dataset as an IEnumerable.
+            IEnumerable<DatasetUtils.SampleTemperatureData> enumerableOfData = DatasetUtils.GetSampleTemperatureData(5);
+
+            // Read in into an IDataView using ReadFromEnumerable. 
+            IDataView data = mlContext.Data.ReadFromEnumerable(enumerableOfData);
+
+            // We can now examine the records in the IDataView. We first create an enumerable of rows in the IDataView.
+            var rowEnumerable = mlContext.Data.CreateEnumerable<DatasetUtils.SampleTemperatureData>(data, true);
```
Comment (wschin) [inline docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/DataViewEnumerable.cs]: Can we also demonstrate the use of the 3rd argument, schemaDefinition? That argument is rarely used and tested.

Comment (artidoro) [reply docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/DataViewEnumerable.cs]: Sounds good will do that!

**Pull request merged**